include/aws/common/ring_buffer.h

```diff
@@ -0,0 +1,81 @@
+#ifndef AWS_COMMON_RING_BUFFER_H
+#define AWS_COMMON_RING_BUFFER_H
+/*
+ * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/common/atomics.h>
+#include <aws/common/byte_buf.h>
+
+/**
+ * Lockless ring buffer implementation that is thread safe assuming a single thread acquires and a single thread
+ * releases. For any other use case (other than the single-threaded use-case), you must manage thread-safety manually.
+ * Also note: release must happen in the same order as acquire.
```

Comment:
We do nothing to enforce this, which could lead to absolutely lawless overwrites. We also don't store the size or a nonce at the head of each buffer, so we can't check to see if release is safe/correctly ordered. Are we ok with that?

Reply:
we don't need size or nonce on head because we're filling in byte buf.

I'm not sure how we would check the ordering with that anyways.

Reply:
also, I don't think we'd be able to have a lockless release at that point.